builder/tart/step_run.go

```diff
@@ -201,41 +201,57 @@ func typeBootCommandOverVNC(
 	dialer := net.Dialer{}
 	netConn, err := dialer.DialContext(ctx, "tcp", fmt.Sprintf("%s:%s", vncHost, vncPort))
 	if err != nil {
-		err := fmt.Errorf("Failed to connect to the Tart's VNC server: %s", err)
+		err := fmt.Errorf("Failed to connect to Tart's VNC server: %s", err)
 		state.Put("error", err)
 		ui.Error(err.Error())
 
 		return false
 	}
 	defer netConn.Close()
 
+	serverMessageChannel := make(chan vnc.ServerMessage)
 	vncClient, err := vnc.Client(netConn, &vnc.ClientConfig{
 		Auth: []vnc.ClientAuth{
 			&vnc.PasswordAuth{Password: vncPassword},
 		},
+		ServerMessageCh: serverMessageChannel,
 	})
 	if err != nil {
-		err := fmt.Errorf("Failed to connect to the Tart's VNC server: %s", err)
+		err := fmt.Errorf("Failed to connect to Tart's VNC server: %s", err)
 		state.Put("error", err)
 		ui.Error(err.Error())
 
 		return false
 	}
 	defer vncClient.Close()
 
-	ui.Say("Connected to the VNC!")
+	ui.Say("Connected to VNC server!")
+
+	vncClient.SetEncodings([]vnc.Encoding{
+		&ZlibEncoding{},
+		&DesktopSizePseudoEncoding{},
+	})
+
+	vncDriver := newCustomDriver(vncClient, serverMessageChannel, config, ctx)
 
 	if config.VNCConfig.BootWait > 0 {
 		message := fmt.Sprintf("Waiting %v after the VM has booted...", config.VNCConfig.BootWait)
 		ui.Say(message)
 		time.Sleep(config.VNCConfig.BootWait)
+	} else {
+		ui.Say(fmt.Sprintf("Waiting for first frame..."))
+		for i := 0; i < 2; i++ {
+			if err := vncDriver.WaitForFramebufferUpdate(); err != nil {
+				state.Put("error", err)
+				ui.Error(err.Error())
+				return false
+			}
+		}
 	}
 
-	message := fmt.Sprintf("Typing commands with key interval %v...", config.BootKeyInterval)
+	message := fmt.Sprintf("Typing commands with key interval %v...", vncDriver.KeyInterval())
```

Comment:
Why resolve key interval manually in `newCustomDriver()` instead of just using the `config.BootKeyInterval`?

Reply:
Because if `config.BootKeyInterval` is not set, `bootcommand.NewVNCDriver` computes the default value via `PackerKeyDefault` and also reads the interval from the `PACKER_KEY_INTERVAL` environment variable. But doesn't have a way to reflect that back. 

Currently, the "Typing commands with key interval" message is wrong if `config.BootKeyInterval` is not set, because we don't pick up if the `PACKER_KEY_INTERVAL` env var is set.